scripts/deterministic/DeployScroll.s.sol

```diff
@@ -0,0 +1,1622 @@
+            notnull(L1_SCROLL_MESSENGER_PROXY_ADDR),
+            notnull(L2_SCROLL_STANDARD_ERC20_ADDR),
+            notnull(L2_SCROLL_STANDARD_ERC20_FACTORY_ADDR)
+        );
+
+        L1_STANDARD_ERC20_GATEWAY_IMPLEMENTATION_ADDR = deploy(
+            "L1_STANDARD_ERC20_GATEWAY_IMPLEMENTATION",
+            type(L1StandardERC20Gateway).creationCode,
+            args
+        );
+
+        upgrade(
+            L1_PROXY_ADMIN_ADDR,
+            L1_STANDARD_ERC20_GATEWAY_PROXY_ADDR,
+            L1_STANDARD_ERC20_GATEWAY_IMPLEMENTATION_ADDR
+        );
+    }
+
+    function deployL1CustomERC20Gateway() private {
+        bytes memory args = abi.encode(
+            notnull(L2_CUSTOM_ERC20_GATEWAY_PROXY_ADDR),
+            notnull(L1_GATEWAY_ROUTER_PROXY_ADDR),
+            notnull(L1_SCROLL_MESSENGER_PROXY_ADDR)
+        );
+
+        L1_CUSTOM_ERC20_GATEWAY_IMPLEMENTATION_ADDR = deploy(
+            "L1_CUSTOM_ERC20_GATEWAY_IMPLEMENTATION",
+            type(L1CustomERC20Gateway).creationCode,
+            args
+        );
+
+        upgrade(L1_PROXY_ADMIN_ADDR, L1_CUSTOM_ERC20_GATEWAY_PROXY_ADDR, L1_CUSTOM_ERC20_GATEWAY_IMPLEMENTATION_ADDR);
+    }
+
+    function deployL1ERC721Gateway() private {
+        bytes memory args = abi.encode(notnull(L2_ERC721_GATEWAY_PROXY_ADDR), notnull(L1_SCROLL_MESSENGER_PROXY_ADDR));
+
+        L1_ERC721_GATEWAY_IMPLEMENTATION_ADDR = deploy(
+            "L1_ERC721_GATEWAY_IMPLEMENTATION",
+            type(L1ERC721Gateway).creationCode,
+            args
+        );
+
+        upgrade(L1_PROXY_ADMIN_ADDR, L1_ERC721_GATEWAY_PROXY_ADDR, L1_ERC721_GATEWAY_IMPLEMENTATION_ADDR);
+    }
+
+    function deployL1ERC1155Gateway() private {
+        bytes memory args = abi.encode(notnull(L2_ERC1155_GATEWAY_PROXY_ADDR), notnull(L1_SCROLL_MESSENGER_PROXY_ADDR));
+
+        L1_ERC1155_GATEWAY_IMPLEMENTATION_ADDR = deploy(
+            "L1_ERC1155_GATEWAY_IMPLEMENTATION",
+            type(L1ERC1155Gateway).creationCode,
+            args
+        );
+
+        upgrade(L1_PROXY_ADMIN_ADDR, L1_ERC1155_GATEWAY_PROXY_ADDR, L1_ERC1155_GATEWAY_IMPLEMENTATION_ADDR);
+    }
+
+    function deployL1GasTokenGateway() private gasToken(true) {
+        bytes memory args = abi.encode(
+            notnull(L1_GAS_TOKEN_ADDR),
+            notnull(L2_ETH_GATEWAY_PROXY_ADDR),
+            notnull(L1_GATEWAY_ROUTER_PROXY_ADDR),
+            notnull(L1_SCROLL_MESSENGER_PROXY_ADDR)
+        );
+
+        L1_GAS_TOKEN_GATEWAY_IMPLEMENTATION_ADDR = deploy(
+            "L1_GAS_TOKEN_GATEWAY_IMPLEMENTATION",
+            type(L1GasTokenGateway).creationCode,
+            args
+        );
+
+        upgrade(L1_PROXY_ADMIN_ADDR, L1_GAS_TOKEN_GATEWAY_PROXY_ADDR, L1_GAS_TOKEN_GATEWAY_IMPLEMENTATION_ADDR);
+    }
+
+    function deployL1WrappedTokenGateway() private gasToken(true) {
+        bytes memory args = abi.encode(notnull(L1_WETH_ADDR), notnull(L1_STANDARD_ERC20_GATEWAY_PROXY_ADDR));
+
+        L1_WRAPPED_TOKEN_GATEWAY_ADDR = deploy(
```

Comment:
Is it intentional that this is not an upgradable proxy contract?

Reply:
I think so, but needs @zimpha to confirm.

Reply:
It's OK according to Xi.